postmortems/2021-07-27_update_ris3.md

```diff
@@ -0,0 +1,67 @@
+
+This last point meant that updating RIS3 using any code-based approach was impossible, because doing so relied on a versioned `ris3_points.ttl` that was out-of-date compared to whatever unversioned `ris3_points.ttl` had been used to generate the existing model data. Using the older copy would have undone newer changes with no way to recover them.
+
+### Update attempt 3
+
+The storypoint was eventually created using the following incredibly brittle method:
+
+- Add the new storypoint into `ris3_views.py` and run this transform only to generate the triples for the storypoint.
+- Open the triples json and find and copy only the new triples.
+- Open the model pulled down from staging and paste the new triples into it, after making sure that `asOrdinal` wasn't going to collide with any existing values.
+- Also edit the markdown for the sidebar in the model to add a link to the new storypoint.
+- Push the model back up to staging.
+
+(This involved a lot of direct editing of a 2.5MB slab of JSON representing the model, which caused Visual Studio Code to crash twice.)
+
+Obviously I have compounded the original problem here by adding even more unversioned data to the model -- although I have made a PR with an update to `ris3_views`, the generated triples will have the wrong tileset version as there's no good way of setting this on a per-storypoint level yet. Even if there was, though, `ris3_views` is still out-of-sync with `ris3_points.ttl`.
+
+### A note on tile generation
+
+Tile generation from PDFs is usually done On Martin's Computer using a shell script. These shell scripts are undocumented; Martin provided one for RIS3 prior to going on holiday but it started with a huge array of commented-out variables (presumably corresponding to the different tilesets for RIS3) that made things generally rather impenetrable. Now that I have gone through this learning process I *think* I could probably use this to generate some tiles, but when looking at it from the other end it was much easier to let Mark N do it as he knew the AI-native method would work.
+
+## Actions
+
+### Immediate (now)
+
+- **Locate the copy of `ris3_points.ttl` was used to generate the existing model on staging and prod and version it.** All code that is used to generate production data should be versioned so that the production data can be reconstructed from scratch if need be.
+
+### Near future (2-4 weeks)
+
+- Backport the manual data edits from `ris3_points.ttl` into `ris3_views.py`.
+- Give `ris3_views.py` the ability to set specific tileset versions per storypoint (may require some additional functionality in `sheet-to-triples`).
+- Once `ris3_points.ttl` can be recreated by running `ris3_views.py`, delete `ris3_points.ttl`.

```

Comment:
Don't think this works. Or, maybe it does but involves writing out a bunch of stuff in a not-any-easier form. The `ris3_points.ttl` content 's taken from data input by the consultants, and is not otherwise present in our transform. There _is_ some useful rationalisation that can happen though. without yet addressing the underlying model problems.

Reply:
Perhaps not-any-easier, but to my mind also not-any-harder? I feel that we should try a bit harder to fit things into our existing model before accepting the existence of special-case files for a map.

The thing I find objectionable about `ris3_points.ttl` is that we have a model of:

- Transforms as `.py` files.
- Which are then added to a `ris3_list.txt`.
- And then, assuming you have the source workbooks and the labels (which is something that we could make more robust separately), generating the data model is a pretty simple command that is structured no differently from the command for generating the highways data.

The workflow here is *more-or-less* the same from map to map. Adding in an extra thing like `ris3_points.ttl` is divergence from that workflow. What if other maps have a `ris3_points.ttl`? What if other maps have more than one? We'd at the very least have to document everything that goes into a map (which we should do anyway) but this process should not be so complicated that I have to remember fifteen different things for fifteen different maps. I know that ultimately `ris3_points.ttl` existing on its own is relatively trivial, I just don't want to normalise it, hence my proposal for killing it.